### PR TITLE
Dedup nomination targets

### PR DESCRIPTION
**Pull request by acgxv**
Deduplicate the targets during the nomination process.

And I'm not pretty sure this is designed to be liked this. 

This was found by the Subscan team. There is only one such special account in Polkadot(need a migration?). The Polkadot Apps UI will filter the selected targets for the users. I assume this uses a custom client to submit the extrinsic.

Note: the targets length affect the weight.

Or should we just return an error?  

---

- https://polkadot.subscan.io/account/138QBy8KBaXZwHXBSmhWALFHhKCmeEnQ8wKZJSwiHx2nhmQ5

---

Polkadot address: 156HGo9setPcU2qhFMVWLkcmtCEGySLwNqa3DaEiYSWtte4Y

### CODE REVIEW

**Comment by kianenigma** [comment]
@gpestana @Ank4n

File: substrate/frame/staking/src/pallet/mod.rs [hunk 1]
```diff
@@ -1286,6 +1286,9 @@ pub mod pallet {
 			}
 
 			ensure!(!targets.is_empty(), Error::<T>::EmptyTargets);
+
+			targets.dedup();
```
Comment (Ank4n) [inline substrate/frame/staking/src/pallet/mod.rs:1290]: I would move this after the `targets.len() <= T::NominationsQuota::get_quota(ledger.active)` check so we are not deduplicating against an unnecessary long vec.

**Review by Ank4n — COMMENTED** [review]
Just one comment; otherwise, lgtm!

We don’t need to migrate the single account. When they update their nomination, any duplicates will be removed automatically. (You could add a unit test to demonstrate that updating nominations removes old duplicates as well.)

Thank you for your PR!

**Review by Ank4n — APPROVED** [review]

File: substrate/frame/staking/src/pallet/mod.rs [hunk 2]
```diff
@@ -1291,6 +1291,8 @@ pub mod pallet {
 				Error::<T>::TooManyTargets
 			);
 
+			targets.dedup();
```
Comment (gui1117) [inline substrate/frame/staking/src/pallet/mod.rs:1294]: This removes only consecutive duplicates, do we enforce it is sorted somewhere?

**Review by kianenigma — CHANGES_REQUESTED** [review]
https://github.com/paritytech/polkadot-sdk/pull/6307/files#r1828951375

File: substrate/frame/staking/src/pallet/mod.rs [hunk 3]
```diff
@@ -1296,21 +1296,26 @@ pub mod pallet {
 				Error::<T>::TooManyTargets
 			);
 
+			let mut targets = targets
+				.into_iter()
+				.map(|t| T::Lookup::lookup(t).map_err(DispatchError::from))
```
Comment (AurevoirXavier) [inline substrate/frame/staking/src/pallet/mod.rs:1301]: Could we use `AccountId` here? I'm not sure why we are using the `Lookup` type in this context.

**Comment by AurevoirXavier** [comment]
@kianenigma @gui1117 🙏

**Comment by aurexav** [comment]
Hi, are we going to merge this?

**Review by gui1117 — APPROVED** [review]
nominate weight is proportional to N, with this PR it is actually `O(n*log(n))` because we are sorting.
(rust is using timsort so worst case is `O(n log n)`.)

Maybe the benchmark should be changed to actually benchmark for the worst case, but I think it is fine as it is.

**Comment by Ank4n** [comment]
> nominate weight is proportional to N, with this PR it is actually `O(n*log(n))` because we are sorting. (rust is using timsort so worst case is `O(n log n)`.)
> 
> Maybe the benchmark should be changed to actually benchmark for the worst case, but I think it is fine as it is.

Since n is a small number (16 for dot, 24 for kusama), I think this is okay.

**Comment by Ank4n** [comment]
@kianenigma Could you review the changes to see if they’re acceptable to you and merge them? They LGTM.

**Comment by aurexav** [comment]
Up...
